### PR TITLE
SwEnt Milestone 1 Feedback

### PR DESCRIPTION
**Pull request by Limandy13**
Use this mechanism to make M1 feedback reports available to teams to whose repos we don't have suitable access

### CODE REVIEW

File: staff-feedback/M1.md [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+
+We analyzed your product backlog, Sprint 3 backlog, and the Scrum board. For the PB, we checked whether it was up-to-date and whether you had populated it with user stories for the first epic and for a second epic. For the Sprint 3 backlog, we looked at whether it was up-to-date and whether the task descriptions were clear. We looked at whether tasks were assigned, whether they had a time estimate, and whether they had a priority. Finally, we evaluated whether your
+Scrum board was up-to-date and ready for Sprint 3.
+
+Sprint 3 backlog was not ready for the meeting
+The description is missing from almost all tasks at the time of the meeting, in the same manner, the task are not assigned, don't have time estimate nor priority.
+
+For this part, you received 8 points out of a maximum of 20.
+
+Based on the above points, your intermediate grade for this milestone M1 is 4.95. If you are interested in how this fits into the bigger grading scheme, please see [project README](https://github.com/swent-epfl/private/blob/main/project/README.md) and the [course README](https://github.com/swent-epfl/public/blob/main/README.md).
```
Comment (ArthurChalard) [inline staff-feedback/M1.md:114]: The link to the project README is in the private repo so we don't have direct access to it.

**Review by ArthurChalard — COMMENTED** [review]
Thank you very much for the quick grading !
I just permitted myself to let you know that there is a broken link below.

**Comment by Plouc314** [comment]
# Milestone M2: Team Feedback

This milestone M2 provides an opportunity to give you, as a team, formal feedback on how you are performing in the project. By now, you should be building upon the foundations set in M1, achieving greater autonomy and collaboration within the team. This is meant to complement the informal, ungraded feedback from your coaches given during the weekly meetings or asynchronously on Discord, email, etc.

The feedback focuses on two major themes:
First, whether you have adopted good software engineering practices and are making progress toward delivering value to your users.
Is your design and implementation of high quality, easy to maintain, and well tested?
Second, we look at how well you are functioning as a team, how you organize yourselves, and how well you have refined your collaborative development.
An important component is also how much you have progressed, as a team, since the previous milestone.
You can find the evaluation criteria in the [M2 Deliverables](https://github.com/swent-epfl/public/blob/main/project/M2.md) document.
As mentioned in the past, the standards for M2 are elevated relative to M1, and this progression will continue into M3.

We looked at several aspects, grouped as follows:

 - Design
   - [Features](#design-features)
   - [Design Documentation](#design-documentation)
 - [Implementation and Delivery](#implementation-and-delivery)
 - Scrum
   - [Backlogs Maintenance](#scrum-backlogs-maintenance)
   - [Documentation and Ceremonies](#scrum-documentation-and-ceremonies)
   - [Continuous Delivery of Value](#scrum-continuous-delivery-of-value)

## Design: Features

We interacted with your app from a user perspective, assessing each implemented feature and flagging any issues encountered. Our evaluation focused mainly on essential features implemented during Sprints 3, 4, and 5; any additional features planned for future Sprints were not considered in this assessment unless they induced buggy behavior in the current APK.
We examined the completeness of each feature in the current version of the app, and how well it aligns with user needs and the overall project goals.


All the features you implemented add substantial value and meet the M2 requirements. You implemented some very complex features.

There are a number of bugs that hinder the user experience and added value of your features:
Adding a new participant to a travel doesn't work, the user is marked as part of the travel, but the user itself doesn't see the travel.
"Unknown" location in activities, first we can not change an unknown location, second an unknown location is placed off the Gulf of Guinea on the map.
The import-file-by-email feature is unstable, from our tests, it took at least 15 minutes for the first file we sent, and the second file we sent has been uploaded twice. Also, this feature is lacking documentation, be it in the app or in the repository.

While most of the features you chose to implement are highly valuable, some are more questionnable, for example having 3 different ways of importing a file. Nonetheless you remain globally focused on your core features.


For this part, you received 6.5 points out of a maximum of 8.0.

## Design: Documentation

We reviewed your Figma (including wireframes and mockups) and the evolution of your overall design architecture in the three Sprints.
We assessed how you leveraged Figma to reason about the UX, ensure a good UX, and facilitate fast UI development.
We evaluated whether your Figma and architecture diagram accurately reflect the current implementation of the app and how well they align with the app's functionality and structure.


The Figma is up to date with the app and show some future screens, however try to make sure the mock up part is consistent with your app design.
The architecture diagram provide a clear overview of the app structure.


For this part, you received 5.4 points out of a maximum of 6.0.

## Implementation and Delivery

We evaluated several aspects of your app's implementation, including code quality, testing, CI practices, and the functionality and quality of the APK.
We assessed whether your code is well modularized, readable, and maintainable.
We looked at the efficiency and effectiveness of your unit and end-to-end tests, and at the line coverage they achieve.


The quality of the code is overall very good, in particular all classes, interfaces and methods in the data layer are very well documented. The MVVM pattern is correctly used and the code is well modularized.
There is still room for improvement: there is commented out code (for example in ui/theme).
Some of the UI code is a bit messy (for example the AddTravelScreen) try to define callback functions instead of writing them in the middle of the UI code.
Also, try using string resources instead of having literal strings in your files, whether it is for test tags, or text that will be displayed to users. Using string resources will allow better maintainability, among other things.
Your app achieve overall a very high coverage, with only one screen below 80% line coverage, however, none of the Firebase cloud function code is tested, you should really try setup tests for it.

The APK is stable, there are some minor UI issues: the buttons in travel settings, the pop up of a participant in participants screen. The overview screen has no bottom bar, no profile button.
One major usability issue, when clicking on the calendar icon in the travel screen, we get stuck in a loading screen.
There is also an interesting choice of colors in the calendar screen that doesn't fit with the rest of the application.
When changing the username of the user, it is displayed "name" instead in the participant screen. We can create activity with date in the past.


For this part, you received 13.2 points out of a maximum of 16.0.

## Scrum: Backlogs Maintenance

We looked at whether your Scrum board is up-to-date and well organized.
We evaluated your capability to organize Sprint 6 and whether you provided a clear overview of this planning on the Scrum board.
We assessed the quality of your user stories and epics: are they clearly defined, are they aligned with a user-centric view of the app, and do they suitably guide you in delivering the highest value possible.


Your sprint backlog was prepared and up to date for sprint 6, however you should make sure that each team member has enough work to do during the sprint.
Your issues are clearly described with acceptance criterias, linked to the relevant user stories.
You should also set the epic your user stories and issues belong to.


For this part, you received 3.6 points out of a maximum of 4.0.

## Scrum: Documentation and Ceremonies

We assessed how you used the Scrum process to organize yourselves efficiently.
We looked at how well you documented your team Retrospective and Stand-Up during each Sprint.
We also evaluated your autonomy in using Scrum.


The scrum documents are filled on time and relevant informations, we can clearly see that they are filled diligently.
The scrum meetings are overall structured and pretty efficiently conducted, with the notable exception of the app demo, please prepare the demo in advance to have it done in one go by one person.
Your team is autonomous in its organization, sprint planning and resolving of blockers. We, the coaches, feel more like stakeholders than coaches (which is very good).
Well done!


For this part, you received 3.9 points out of a maximum of 4.0.

## Scrum: Continuous Delivery of Value

We evaluated the Increment you delivered at the end of each Sprint, assessing your team’s ability to continuously add value to the app.
This included an assessment of whether the way you organized the Sprints was conducive to an optimal balance between effort invested and delivery of value.


You have been able to deliver value at each sprint, even if it was not perfectly consistent, with some sprint adding big features and the last one at little bit lighter polishing the app.


For this part, you received 1.8 points out of a maximum of 2.0.

## Summary

Based on the above points, your intermediate grade for this milestone M2 is 5.30. If you are interested in how this fits into the bigger grading scheme, please see the [project README](https://github.com/swent-epfl/public/blob/main/project/README.md) and the [course README](https://github.com/swent-epfl/public/blob/main/README.md).

Your coaches will be happy to discuss the above feedback in more detail.

Good luck for the next Sprints!

**Comment by Plouc314** [comment]
# Milestone M3: Team Feedback

This milestone M3 is the culmination of your SwEnt journey, and it gives us the final opportunity to give you, as a team, formal feedback on how you performed in the project. By now, you should be capable of demonstrating a solid command of the Scrum methodology and collaborative teamwork, and be able to deliver a high-quality, application that is ready for real users.
This feedback report is meant to complement the informal, ungraded feedback that you received from your coaches during the weekly meetings, over email, on Discord, etc.

You can find the evaluation criteria in the [M3 Deliverables](https://github.com/swent-epfl/public/blob/main/project/M3.md) document.
As mentioned before, the standards for M2 were elevated relative to M1, and this progression continued into M3: we now hold you to the highest professional standard in SwEnt.

For this milestone, we looked at several aspects, grouped as follows:

- Application
  - [Completeness](#app-completeness)
  - [Functionality](#app-functionality)
  - [User Experience](#app-user-experience)
- [Design and Implementation](#design-and-implementation)
- [Testing and CI Pipeline](#testing-and-ci-pipeline)
- Process
  - [Documentation](#process-documentation)
  - [Autonomy](#process-autonomy)

## App: Completeness

We first evaluated the depth and complexity of the main __epics__ in your app, along with their contribution to the app, the tangible value they provide to the user, and their alignment with the app’s goals.
We evaluated the extent to which your app meets the __course requirements__ articulated at the start of the semester, and whether they are implemented effectively, they integrate seamlessly, and are indeed essential to the app.
We then looked at the __robustness and completeness__ of the different features you implemented: are all the features finished and polished, are they secure and bug-free, and are they thoughtfully designed.


Your app fully implements multiple epics, they are really cool and exceed the expectations.

The app meets all the course requirements, are well integrated and show advance usage, especially the offline mode and the sensors: both the GPS and the camera add substantial value to the app. Multiple cloud service capabilities are used, notably the cloud functions, well done!

All the features you developped are complete and polished (except the storage dashboard which still has hardcoded values).


For this part, you received 7.8 points out of a maximum of 8.0.

## App: Functionality

In this context, we assessed your app's ability to __handle unexpected inputs__ provided by clueless or malicious users (including spamming buttons, entering wrong inputs, stopping a process mid-way, etc.); we wanted to see that your app handles all edge cases gracefully, has comprehensive error handling, and includes robust mechanisms for maintaining stability under stress.

We then evaluated the performance and reliability of the final product, i.e., __the APK__: we wanted to see that your APK is stable and delivers excellent performance, the UI responds quickly and has seamless navigation.

Next we looked into your implementation of __user authentication and multi-user support__: does the app correctly manage users, can users personalize their accounts, does the app support session persistence, are multi-user interactions well supported, can a user recover a lost password, can accounts be used on another device, and is account information preserved when switching devices.


The app is robust, note that you can add activities whose date does not fall within the travel.

The APK is stable and performant, the UI is smooth and responsive. It works on multiple type of devices, including tablets.

The user can create multiple accounts, from multiple authentication providers, it supports multi user interactions, and has session persistence, good job!


For this part, you received 7.6 points out of a maximum of 8.0.

## App: User Experience

For this part, we wanted to see how __intuitive and user-friendly__ the app is for real users. Beyond having good usability, did you pay attention to streamlining the interactions, is it easy to figure out, can new users start making good use of the app quickly, are the interaction flows well thought out and refined.


The application is globally intuitive to use, with notably an onboarding flow, and some indication (for creating a new travel), however the number of features is such that it introduces a learning curve.


For this part, you received 2 points out of a maximum of 2.0.

## Design and Implementation

We evaluated whether __your code is of high quality and employs best practices__, is it modular and robust, etc.
We expect the codebase to be polished, well documented, follow consistent conventions, be modular, and allow for easy modifications.
You should be able to employ advanced techniques by now, such as asynchronous functions (flows, coroutines), good resource management, and automated dependency injection (e.g., with Hilt).

We assessed your overall __app architecture and design__, looking in particular at aspects surrounding robustness and scalability.
We looked at both the codebase and the documentation of the app (Wiki and architecture diagram).
We expect your design to demonstrate thoughtful consideration for performance, maintainability, and future growth.


The codebase is vely clean, it includes the use of advance techniques, notably dependency injection (Hilt), cloud functions and push notifications. It is very well documented.

The architecture is well thought, with some clever ideas such as the use of an existing folder to cache the documents.


For this part, you received 8 points out of a maximum of 8.0.

## Testing and CI Pipeline

The first aspect we looked at here was your __test suite__, in terms of both quality and the final line coverage.
We expect testing the be rigorous and to cover all components and edge cases, and they should validate every significant user journey.
Line coverage should be getting close to 90%.
Your end-to-end tests should be detailed and include error-handling scenarios.
The tests should be well-documented and easy to maintain.
Finally, your test suite should demonstrate  advanced techniques, mock data for performance testing, and automated regression tests.

We then considered the quality of your __repository setup and the CI pipeline__, and how professional it is and how easy it is for new developers to bring contributions to the project.
We expect your repository to have a logical structure, use consistent naming, and take full advantage of CI (code quality checks, linting, formatting, etc.)
Ideally, you should also have automated performance testing, deployment pipelines, and the CI should provide detailed feedback for developers.


The project contains 4 end-to-end tests and achieves 88% line coverage over the whole codebase, well done!

The repository is professional, the CI is setup to build the APK on trigger additionnaly to running tests.


For this part, you received 7.2 points out of a maximum of 8.0.

## Process: Documentation

We looked at your `README` and GitHub Wiki to evaluate the quality and completeness of __your app’s documentation__. We expect the README and Wiki to be thorough and achieve professional-level clarity and completeness.
They should provide detailed descriptions of the app's architecture, implementation of the features, and the development setup.
We also assessed __your use of Figma and the architecture diagram__ for effective UI design, organization, and app structure planning.
By this stage, we expect your Figma to be complete, up-to-date, and to include UI elements for future features, showing foresight and organization.
The architecture diagram should be comprehensive, providing a clear and detailed overview of the app structure and dependencies.
The architecture should be robust, scalable, and optimized for performance, security, and future development.


Your README contains the most important informations and relevant links.

The Figma doesn't seem to be up to date, the mockup doesn't match the app current UI. The architecture diagram is simplistic.


For this part, you received 3 points out of a maximum of 4.0.

## Process: Autonomy

A primary goal of SwEnt is to teach you how to __function autonomously as a team__.
For this part of the evaluation, we assessed you team’s independence, spanning Sprint 6 to Sprint 10, based on the meetings with coaches, Sprint planning, and how you managed risk.
By this stage, coaches should no longer be necessary for the team to operate, i.e., you can organize yourselves, you don't need to be reminded about tasks, and you can conduct the Scrum ceremonies on your own.


Your team is autonomous and we don't need to intervene anymore.


For this part, you received 2 points out of a maximum of 2.0.

## Summary

Based on the above points, your intermediate grade for this milestone M3 is 5.70.

The entire SwEnt staff wishes you the very best in your career, and we look forward to seeing you do great things with what you learned this semester.
